File: 2025/day7/main.py
with open("input.txt", "r") as f:
    grid = [line.rstrip("\n") for line in f]

# Find S
start_row, start_col = None, None
for r in range(len(grid)):
    for c in range(len(grid[r])):
        if grid[r][c] == "S":
            start_row, start_col = r, c
            break

# Active beams: set of (row, col)
beams = {(start_row, start_col)}
split_count = 0

while beams:
    new_beams = set()

    for row, col in beams:
        next_row = row + 1

        # Check if out of bounds (bottom)
        if next_row >= len(grid):
            continue

        # Check if out of bounds (left/right)
        if col < 0 or col >= len(grid[next_row]):
            continue

        # Check what's at the next position
        if grid[next_row][col] == "^":
            # Hit a splitter!
            split_count += 1
            # Create two new beams at left and right
            new_beams.add((next_row, col - 1))
            new_beams.add((next_row, col + 1))
        else:  # '.' or 'S'
            # Continue downward
            new_beams.add((next_row, col))

    beams = new_beams

print(split_count)

File: 2025/day7/main.go
package main

import (
	"bufio"
	"fmt"
	"os"
)

type Pos struct {
	row, col int
}

func main() {
	file, _ := os.Open("input.txt")
	defer file.Close()

	var grid []string
	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		grid = append(grid, scanner.Text())
	}

	// Find S
	var startRow, startCol int
	for r, line := range grid {
		for c, ch := range line {
			if ch == 'S' {
				startRow, startCol = r, c
				break
			}
		}
	}

	// Active beams
	beams := make(map[Pos]bool)
	beams[Pos{startRow, startCol}] = true
	splitCount := 0

	for len(beams) > 0 {
		newBeams := make(map[Pos]bool)

		for pos := range beams {
			nextRow := pos.row + 1

			// Check bounds
			if nextRow >= len(grid) {
				continue
			}
			if pos.col < 0 || pos.col >= len(grid[nextRow]) {
				continue
			}

			// Check next position
			if grid[nextRow][pos.col] == '^' {
				// Split!
				splitCount++
				newBeams[Pos{nextRow, pos.col - 1}] = true
				newBeams[Pos{nextRow, pos.col + 1}] = true
			} else {
				// Continue down
				newBeams[Pos{nextRow, pos.col}] = true
			}
		}

		beams = newBeams
	}

	fmt.Println(splitCount)
}
